refactor!(chain): Unify ChangeSet nomenclature

danielabrozzoni opened the pull request:
Solves #1022

### Changelog notice

Rename `indexed_tx_graph::IndexedAdditions` to `indexed_tx_graph::ChangeSet`
Rename `indexed_tx_graph::IndexedAdditions::graph_additions` to `indexed_tx_graph::ChangeSet::graph`
Rename `indexed_tx_graph::IndexedAdditions::index_additions` to `indexed_tx_graph::ChangeSet::indexer`
Rename `tx_graph::Additions` to `tx_graph::ChangeSet`
Rename `keychain::DerivationAdditions` to `keychain::ChangeSet`
Rename `CanonicalTx::node` to `CanonicalTx::tx_node`
Rename `CanonicalTx::observed_as` to `CanonicalTx::chain_position`
Rename `LocalChangeSet` to `WalletChangeSet`
Rename `LocalChangeSet::chain_changeset` to `WalletChangeSet::chain`
Rename `LocalChangeSet::indexed_additions` to `WalletChangeSet::indexed_tx_graph`
Rename `LocalUpdate` to `WalletUpdate`
Make `TxGraph::determine_changeset` `pub(crate)`
Add `TxGraph::initial_changeset`
Add `IndexedTxGraph::initial_changeset`
Remove `TxGraph::insert_txout_preview`
Remove `TxGraph::insert_tx_preview`
Remove `insert_anchor_preview`
Remove `insert_seen_at_preview`

### Notes to reviewers

I'm not sure whether we want to keep `TxGraph::determine_changeset` public or not

### Checklists

#### All Submissions:

* [x] I've signed all my commits
* [x] I followed the [contribution guidelines](https://github.com/bitcoindevkit/bdk/blob/master/CONTRIBUTING.md)
* [x] I ran `cargo fmt` and `cargo clippy` before committing

danielabrozzoni committed:
refactor!(chain): Unify ChangeSet nomenclature

This commit renames:

- indexed_tx_graph::IndexedAdditions -> indexed_tx_graph::ChangeSet
- indexed_tx_graph::IndexedAdditions::graph_additions -> indexed_tx_graph::ChangeSet::graph
- indexed_tx_graph::IndexedAdditions::index_additions -> indexed_tx_graph::ChangeSet::indexer
- tx_graph::Additions -> tx_graph::ChangeSet
- keychain::DerivationAdditions -> keychain::ChangeSet
- CanonicalTx::node -> CanonicalTx::tx_node
- CanonicalTx::observed_as -> CanonicalTx::chain_position
- LocalChangeSet -> WalletChangeSet
- LocalChangeSet::chain_changeset -> WalletChangeSet::chain
- LocalChangeSet::indexed_additions -> WalletChangeSet::indexed_tx_graph
- LocalUpdate -> WalletUpdate

This commit also changes the visibility of TxGraph::determine_changeset
to be pub(crate), and the method accepts a TxGraph instead of &TxGraph

This commit removes:
- `TxGraph::insert_txout_preview`
- `TxGraph::insert_tx_preview`
- `insert_anchor_preview`
- `insert_seen_at_preview`

Solves #1022

notmandatory commented on crates/chain/src/tx_graph.rs:
```diff
@@ -1026,7 +984,7 @@ impl<A: Anchor> TxGraph<A> {
 /// A structure that represents changes to a [`TxGraph`].
 ///
-/// It is named "additions" because [`TxGraph`] is monotone, so transactions can only be added and
+/// It is named "changeset" because [`TxGraph`] is monotone, so transactions can only be added and
```
This line makes a less sense with the name change. How about something like:
```suggestion
/// Because [`TxGraph`] is monotone "changeset" can only contain transactions to be added and
```

notmandatory commented on crates/chain/tests/test_tx_graph.rs:
```diff
@@ -135,20 +135,20 @@ fn insert_txouts() {
 
-    // Apply addition and check the new graph counts.
-    graph.apply_additions(additions);
+    // Check the new graph counts.
```
Doesn't this still apply the changeset?
```suggestion
    // Apply changeset and check the new graph counts.
```

notmandatory reviewed (commented):
LGTM! much easier to read using the same nomenclature across the code base. I added a couple comments/questions. I think this should go in before #1048 and then I'll rebase that PR on these changes.

evanlinjin commented:
LGTM too. I can ACK and merge after wording changes as mentioned by @notmandatory.

LLFourn commented:
Haven't reviewed but got a note here: https://github.com/bitcoindevkit/bdk/issues/1060#issuecomment-1669237763 that `TxGraph` doesn't have an "initial_changeset" method to dump the full state as a changeset. It should in order to be consistent. Same with `IndexedTxGraph`. Can we also make sure all changeset types are `must_use` (except for the wallet one).

LLFourn commented on crates/chain/src/indexed_tx_graph.rs:
```diff
@@ -181,64 +181,64 @@ where
+    /// [`TxGraph`] changeset.
+    pub graph_changeset: tx_graph::ChangeSet<A>,
+    /// [`Indexer`] changeset.
+    pub index_changeset: IA,
```
It's inside a thing that's a changeset so I'd rather remove this suffix.

```suggestion
    pub graph: tx_graph::ChangeSet<A>,
    /// [`Indexer`] changeset.
    pub indexer: IA,
```

LLFourn commented on crates/chain/src/keychain.rs:
```diff
@@ -132,29 +129,29 @@ pub struct LocalChangeSet<K, A> {
     ///
     /// [`IndexedTxGraph`]: crate::indexed_tx_graph::IndexedTxGraph
-    pub indexed_additions: IndexedAdditions<A, DerivationAdditions<K>>,
+    pub indexed_changeset: indexed_tx_graph::ChangeSet<A, ChangeSet<K>>,
```
I suggest renaming the following:

- `chain_changeset` field here to just `chain`
- `indexed_changeset` field here to `indexed_tx_graph` 
- Rename `LocalChangeSet` to `WalletChangeSet`
- Rename `LocalUpdate` to `WalletUpdate`. 

Those last two are suggestions would set the stage a bit for https://github.com/bitcoindevkit/bdk/issues/1061 (I want to move both those types to `bdk/wallet/mod.rs`.

LLFourn reviewed (approved):
Approved as it is but I've made a few extra renaming suggestions and I do think now is the time to create `initial_changeset `methods for the tx graphs. The tx graph one could probably be implemented something like:

```rust
fn initial_changeset(&self) -> ChangeSet {
    Self::default().apply_update(self.clone())
}
```

but for `IndexedTxGraph` you might have to add a `initial_changeset` method to the `Indexer` trait to get it done properly.

danielabrozzoni commented:
I rebased and fixed the review comments. I'm now working on adding `initial_changeset` for the `tx_graph` and the `indexed_tx_graph`, but since this PR is holding up #1040, feel free to merge before I'm done and I'll open a new PR

danielabrozzoni commented:
Also, every ChangeSet is `#[must_use]` apart from the local_chain one, but since it's just a type alias, when I tried to add it the compiler warned me that it would have no effect:

```
   Compiling bdk_chain v0.5.0 (/home/daniela/Developer/bitcoindevkit/bdk/crates/chain)
warning: `#[must_use]` has no effect when applied to a type alias
  --> crates/chain/src/local_chain.rs:14:1
   |
14 | #[must_use]
   | ^^^^^^^^^^^
   |
   = note: `#[warn(unused_attributes)]` on by default
```

danielabrozzoni committed:
feat(chain): Add `initial_changeset` to graphs

Add `initial_changeset` to TxGraph and IndexedTxGraph

danielabrozzoni commented:
Ok, this took less than what I expected :) I just pushed 4ac9c03fc5552026460ded3c1d0b9cea87e01243, adding `initial_changeset` to `tx_graph` and `indexed_tx_graph`.

notmandatory reviewed (approved):
Looks good. ACK 4ac9c03fc5552026460ded3c1d0b9cea87e01243

vladimirfomene commented on crates/chain/src/tx_graph.rs:
```diff
@@ -389,100 +385,62 @@ impl<A: Clone + Ord> TxGraph<A> {
-        let additions = self.insert_txout_preview(outpoint, txout);
-        self.apply_additions(additions.clone());
-        additions
+        let changeset = self.determine_changeset(&update);
```
I see a recurrent pattern in these TxGraph methods of calling `determine_changeset` and then `apply_changeset`. Can't we just replace these with `apply_update`?

danielabrozzoni replied:
I think we could, but it's better if we do it in another PR, this one is pretty big already :)

vladimirfomene commented on crates/chain/src/indexed_tx_graph.rs:
```diff
@@ -61,6 +61,13 @@ impl<A: Anchor, I: Indexer> IndexedTxGraph<A, I> {
         self.graph.apply_changeset(graph);
     }
+
+    /// Determines the changeset between self and the empty [`IndexedTxGraph`].
```
Do we not need to write self as `self`?

evanlinjin replied:
Nit: This is more descriptive of how this method should be used.

```suggestion
    /// Determines the [`ChangeSet`] between `self` and an empty [`IndexedTxGraph`].
    ///
    /// The returned `ChangeSet` can be applied to an empty `IndexedTxGraph` to recover the current
    /// state of `self`.
```

vladimirfomene commented on crates/chain/src/indexed_tx_graph.rs:
```diff
@@ -234,6 +241,9 @@ pub trait Indexer {
     /// Apply changeset to itself.
     fn apply_changeset(&mut self, changeset: Self::ChangeSet);
 
+    /// Determines the changeset between self and the empty [`Indexer`].
```
Same as above

vladimirfomene commented on crates/chain/src/tx_graph.rs:
```diff
@@ -438,6 +438,11 @@ impl<A: Clone + Ord> TxGraph<A> {
         changeset
     }
 
+    /// Determines the changeset between self and the empty [`TxGraph`].
```
should we not also put self here in ``?

vladimirfomene reviewed (commented):
Review ACK 4ac9c03fc5552026460ded3c1d0b9cea87e01243. Feel free to push the TxGraph refactoring to another PR.

evanlinjin commented on crates/chain/src/indexed_tx_graph.rs:
```diff
@@ -46,56 +46,57 @@ impl<A, I> IndexedTxGraph<A, I> {
-        } = additions;
+    /// Applies the [`ChangeSet`] to the [`IndexedTxGraph`].
+    pub fn apply_changeset(&mut self, changeset: ChangeSet<A, I::ChangeSet>) {
+        let ChangeSet { graph, indexer } = changeset;
```
Nit: Since we changed the fields of `indexed_tx_graph::ChangeSet`, variable names are now confusing. I think using `changeset.graph` and `changeset.indexer` directly will be better.

evanlinjin commented on crates/chain/src/tx_graph.rs:
```diff
@@ -528,36 +491,36 @@ impl<A: Clone + Ord> TxGraph<A> {
     /// exist in `update` but not in `self`).
-    pub fn determine_additions(&self, update: &TxGraph<A>) -> Additions<A> {
-        let mut additions = Additions::default();
+    pub(crate) fn determine_changeset(&self, update: &TxGraph<A>) -> ChangeSet<A> {
```
Nit: I would also make `determine_changeset` consume `update`.

```suggestion
    pub(crate) fn determine_changeset(&self, update: TxGraph<A>) -> ChangeSet<A> {
```

evanlinjin commented on crates/chain/src/tx_graph.rs:
```diff
@@ -389,100 +385,67 @@ impl<A: Clone + Ord> TxGraph<A> {
+
+    /// Determines the changeset between self and the empty [`TxGraph`].
+    pub fn initial_changeset(&self) -> ChangeSet<A> {
+        Self::default().apply_update(self.clone())
```
Nit: This might be more efficient as we don't need to apply real changes to `Self::default()`.

```suggestion
        Self::default().determine_changeset(self)
```

evanlinjin reviewed (approved):
ACK 4ac9c03fc5552026460ded3c1d0b9cea87e01243

Looks good. My comments are regarding documentation, variable naming and minute implementation details (things that are not mission-critical and can be addressed later).

notmandatory commented:
@danielabrozzoni looks like some small but useful suggestions so I'll wait until you've had a chance to address them before merging.

LLFourn commented:
ACK 4ac9c03fc5552026460ded3c1d0b9cea87e01243

danielabrozzoni committed:
ci: Update tokio version

danielabrozzoni commented:
Thanks for the reviews! Updated, and pushed a commit on top to fix the CI

evanlinjin reviewed (approved):
ACK 62f253103c3fd158ad66a0bba11399e7083226f2
